src/rest_framework_jwt/utils.py

```diff
@@ -107,17 +121,54 @@ def jwt_decode_token(token):
     options = {
         'verify_exp': api_settings.JWT_VERIFY_EXPIRATION,
     }
-    # get user from token, BEFORE verification, to get user secret key
-    unverified_payload = jwt.decode(token, None, False)
-    secret_key = jwt_get_secret_key(unverified_payload)
-    return jwt.decode(
-        token, api_settings.JWT_PUBLIC_KEY or secret_key,
-        api_settings.JWT_VERIFY, options=options,
-        leeway=api_settings.JWT_LEEWAY, audience=api_settings.JWT_AUDIENCE,
-        issuer=api_settings.JWT_ISSUER, algorithms=[
-            api_settings.JWT_ALGORITHM
-        ]
-    )
+
+    algos = api_settings.JWT_ALGORITHM
+    if not isinstance(algos, list):
+        algos = [algos]
+
+    hdr = jwt.get_unverified_header(token)
+    alg = hdr["alg"]
+    if alg not in algos:
+        raise jwt.exceptions.InvalidAlgorithmError
+
+    kid = hdr["kid"] if "kid" in hdr else None
+
+    keys = None
+    if alg.startswith("HS"):
+        unverified_payload = jwt.decode(token, None, False)
+        keys = jwt_get_secret_key(unverified_payload)
+    else:
+        keys = api_settings.JWT_PUBLIC_KEY
+
+    # if keys are named and the jwt has a kid, only consider exactly that key
+    # otherwise if the JWT has no kid, JWT_INSIST_ON_KID selects if we fail
+    # or try all defined keys
+    if isinstance(keys, dict):
+        if kid:
+            try:
+                keys = keys[kid]
+            except:
```

Comment:
I assume this can only raise a `KeyError`, but the current implementation catches everything, including `KeyboardInterrupt`s. I recommend that the exceptions are explicitly stated.

Reply:
right